### ISSUE TITLE
[Bug] Move Prover proves wrong spec for shifting operator

### ISSUE THREAD

**Issue by whst** [diem/move#147]
# 🐛 Bug

Move prover proves  wrong spec for shifting operator.

## To reproduce

Move prover can successfully find the verification error for the following module:
```rust
module 0x1234::Test {
    fun shift(x: u64): u64 { x << 8 }
    spec shift { ensures result == x * 256 + 1; }  // The +1 should cause an error
}
```
But it cannot find the problem for this almost identical version:
```rust
module 0x1234::Test {
    fun shift(x: u64): u64 { x << 7 }
    spec shift { ensures result == x * 128 + 1; }  // The +1 should cause an error
}
```

## Expected Behavior

MVP should report "post-condition does not hold" for both version of the code.

## Additional context

After some investigation, I found that in `move-prover/boggie-backend/src/prelude/prelude.bpl`, `$shl` and `$shr` are defined to support shifting by 8, 16, 32, and 64-bits only. Is that feature on purporse or a bug?

**Comment by DavidLDill** [comment]
I've submitted a PR for this bug, which I hope to be able to land very soon.

Do you expect to be using the Move Prover more in the near future? Are you affiliated with one of the blockchain projects using Move?

Thanks,
Dave
________________________________
From: 五好蒜苔WHsT ***@***.***>
Sent: Thursday, March 24, 2022 5:55 PM
To: diem/move ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [diem/move] [Bug] Move Prover proves wrong spec for shifting operator (Issue #147)


🐛 Bug

Move prover proves wrong spec for shifting operator.

To reproduce

Move prover can successfully find that the verification error for the following module:

module 0x1234::Test {

    fun shift(x: u64): u64 { x << 256 }

    spec shift { ensures result == x * 256 + 1; }  // The +1 should cause an error

}

But it cannot find the problem for this almost identical version:

module 0x1234::Test {

    fun shift(x: u64): u64 { x << 7 }

    spec shift { ensures result == x * 128 + 1; }  // The +1 should cause an error

}

Expected Behavior

MVP should report "post-condition does not hold" for both version of the code.

Additional context

After some investigation, I found that in move-prover/boggie-backend/src/prelude/prelude.bpl, $shl and $shr are defined to support shifting by 8, 16, 32, and 64-bits only. Is that feature on purporse or a bug?

**Comment by whst** [comment]
Yes, I am trying to prove the correctness of the stdlib in starcoin project.

**Comment by DavidLDill** [comment]
This bug should be fixed by PR #155, which was just merged.
